src/backend/distributed/planner/distributed_planner.c

```diff
@@ -527,14 +543,44 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
 
 
 /*
- * CreateDistributedPlannedStmt encapsulates the logic needed to transform a particular
- * query into a distributed plan that is encapsulated by a PlannedStmt.
+ * CreateDistributedPlannedStmt encapsulates the logic needed to transform
+ * a particular query into a distributed plan that is encapsulated by a
+ * PlannedStmt.
  */
 static PlannedStmt *
 CreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan, Query *originalQuery,
 							 Query *query, ParamListInfo boundParams,
 							 PlannerRestrictionContext *plannerRestrictionContext)
 {
+	PlannedStmt *result = NULL;
+
+	if (QueryTreeContainsInlinableCTE(originalQuery))
+	{
+		/*
+		 * Inlining CTEs as subqueries in the query can avoid recursively
+		 * planning some (or all) of the CTEs. In other words, the inlined
+		 * CTEs could become part of query pushdown planning, which is much
+		 * more efficient than recursively planning. So, first try distributed
+		 * planning on the inlined CTEs in the query tree.
+		 *
+		 * We also should fallback to distributed planning with non-inlined CTEs
+		 * if the distributed planning fails with inlined CTEs, because recursively
+		 * planning CTEs can provide full SQL coverage, although it might be slow.
+		 */
+		result = InlineCtesAndCreateDistributedPlannedStmt(planId, localPlan,
+														   originalQuery,
+														   query, boundParams,
+														   plannerRestrictionContext);
+		if (result != NULL)
```

Comment:
might be nicer to return a DistributedPlan with a planning error (the one that was caught)